CA1034 (Nested types should not be visible) reported for delegate declaration

#### Analyzer package

Microsoft.CodeQuality.Analyzers

#### Analyzer

NestedTypesShouldNotBeVisibleAnalyzer

#### Repro steps

```
public class A
{
    public delegate void B();
}
```

#### Expected behavior
No CA1034

#### Actual behavior
`warning CA1034: Do not nest type B. Alternatively, change its accessibility so that it is not externally visible.
`

--- a/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/NestedTypesShouldNotBeVisible.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/NestedTypesShouldNotBeVisible.cs
@@ -76,6 +76,12 @@ namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
                                 return;
                             }
 
+                            // Do not report diagnostic for compiler generated nested types for delegate declaration
+                            if (nestedType.TypeKind == TypeKind.Delegate)
+                            {
+                                return;
+                            }
+
                             // The Framework Design Guidelines (see 4.9 Nested Types) say that it is okay
                             // to expose nested types for advanced customization and subclassing scenarios,
                             // so, following FxCop's implementation of this rule, we allow protected and

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/NestedTypesShouldNotBeVisibleTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/NestedTypesShouldNotBeVisibleTests.cs
@@ -361,6 +361,30 @@ End Class
             VerifyBasic(code);
         }
 
+        [Fact, WorkItem(1347, "https://github.com/dotnet/roslyn-analyzers/issues/1347")]
+        public void CSharpNoDiagnosticDelegateTypeSpecialCases()
+        {
+            var code = @"
+public class A
+{
+    public delegate void B();
+}
+";
+            VerifyCSharp(code);
+        }
+
+        [Fact, WorkItem(1347, "https://github.com/dotnet/roslyn-analyzers/issues/1347")]
+        public void BasicNoDiagnosticDelegateTypeSpecialCases()
+        {
+            var code = @"
+Public Class A
+	Public Delegate Sub B()
+End Class
+";
+            VerifyBasic(code);
+        }
+
+
         [Fact]
         public void CSharpDiagnosticDataSetWithOtherNestedClass()
         {

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/NestedTypesShouldNotBeVisibleTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/NestedTypesShouldNotBeVisibleTests.cs
@@ -378,12 +378,11 @@ public class A
         {
             var code = @"
 Public Class A
-	Public Delegate Sub B()
+    Public Delegate Sub B()
 End Class
 ";
             VerifyBasic(code);
         }
-
 
         [Fact]
         public void CSharpDiagnosticDataSetWithOtherNestedClass()

--- a/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/NestedTypesShouldNotBeVisibleTests.cs
+++ b/src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/NestedTypesShouldNotBeVisibleTests.cs
@@ -100,7 +100,7 @@ End Module
             VerifyBasic(code, GetBasicCA1034ModuleResultAt(3, 18, "Inner"));
         }
 
-        [Fact]
+        [Fact, WorkItem(1347, "https://github.com/dotnet/roslyn-analyzers/issues/1347")]
         public void CSharpDiagnosticPublicNestedDelegate()
         {
             var code = @"
@@ -109,10 +109,10 @@ public class Outer
     public delegate void Inner();
 }
 ";
-            VerifyCSharp(code, GetCSharpCA1034ResultAt(4, 26, "Inner"));
+            VerifyCSharp(code);
         }
 
-        [Fact]
+        [Fact, WorkItem(1347, "https://github.com/dotnet/roslyn-analyzers/issues/1347")]
         public void BasicDiagnosticPublicNestedDelegate()
         {
             var code = @"
@@ -120,7 +120,7 @@ Public Class Outer
     Delegate Sub Inner()
 End Class
 ";
-            VerifyBasic(code, GetBasicCA1034ResultAt(3, 18, "Inner"));
+            VerifyBasic(code);
         }
 
         [Fact]
@@ -356,29 +356,6 @@ Public Class MyDataSet
             MyBase.New(builder)
         End Sub
     End Class
-End Class
-";
-            VerifyBasic(code);
-        }
-
-        [Fact, WorkItem(1347, "https://github.com/dotnet/roslyn-analyzers/issues/1347")]
-        public void CSharpNoDiagnosticDelegateTypeSpecialCases()
-        {
-            var code = @"
-public class A
-{
-    public delegate void B();
-}
-";
-            VerifyCSharp(code);
-        }
-
-        [Fact, WorkItem(1347, "https://github.com/dotnet/roslyn-analyzers/issues/1347")]
-        public void BasicNoDiagnosticDelegateTypeSpecialCases()
-        {
-            var code = @"
-Public Class A
-    Public Delegate Sub B()
 End Class
 ";
             VerifyBasic(code);

